### PR TITLE
improve mutex locking when writing index in boltdb-shipper

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
We currently take an exclusive lock for each index write to check whether desired index file exists which can cause contention. This PR improves the code by checking index file exists with a read lock and then takes a write lock to create if it does not exist.

### CODE REVIEW

**Comment by liguozhong** [comment]
The code of this PR would be helpful. After the PR is merged, we will release boltdb-shipper again to observe the optimization effect.

**Comment by liguozhong** [comment]
I am waiting for this PR to be merged, please help to speed up the merge of this PR.

**Comment by liguozhong** [comment]
@sandeepsukhani Thanks for the quick response

**Comment by liguozhong** [comment]
It would be great if this PR could be merged in the 2.5 release.

**Comment by liguozhong** [comment]
@cyriltovena Please help to review this PR as soon as possible，thanks!

**Review by dannykopping — APPROVED** [review]
LGTM

**Comment by sandeepsukhani** [comment]
cla stuck, will see if reopening PR fixes it.

File: pkg/storage/stores/shipper/uploads/table.go
Comment (bboreham) [inline pkg/storage/stores/shipper/uploads/table.go:242]: I might invert this `if` so that the read-lock and write-lock cases look the same.  Also this would outdent the remainder of the function which is more idiomatic Go.

File: pkg/storage/stores/shipper/uploads/table.go
Comment (bboreham) [inline pkg/storage/stores/shipper/uploads/table.go:240]: This could be moved inside the scope where it is used.

File: pkg/storage/stores/shipper/uploads/table.go
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/uploads/table.go:242]: Sorry, I merged the PR already. Will open a follow-up PR.

**Pull request merged**
